python/ray/ml/train/integrations/huggingface/huggingface_trainer.py

```diff
@@ -146,52 +220,297 @@ def trainer_init_per_worker(train_dataset, eval_dataset, **config):
+                    if get_session() is None:
+                        return super().get_train_dataloader()
+                    return DataLoader(
+                        self.train_dataset,
+                        batch_size=self.args.per_device_train_batch_size,
+                        collate_fn=self.data_collator,
+                        num_workers=self.args.dataloader_num_workers,
+                        pin_memory=self.args.dataloader_pin_memory,
+                    )
+
+                def _wrap_model(self, model, training=True):
+                    if get_session() is None:
+                        return super()._wrap_model(model, training=training)
+
+                    if not training:
+                        return model
+                    kwargs = {}
+                    # same logic as in transformers.Trainer
+                    if self.args.ddp_find_unused_parameters is not None:
+                        kwargs[
+                            "find_unused_parameters"
+                        ] = self.args.ddp_find_unused_parameters
+                    elif isinstance(model, transformers.trainer.PreTrainedModel):
+                        # find_unused_parameters breaks checkpointing as per
+                        # https://github.com/huggingface/transformers/pull/4659#issuecomment-643356021
+                        kwargs[
+                            "find_unused_parameters"
+                        ] = not model.is_gradient_checkpointing
+                    else:
+                        kwargs["find_unused_parameters"] = True
+
+                    if self.args.ddp_bucket_cap_mb is not None:
+                        kwargs["bucket_cap_mb"] = self.args.ddp_bucket_cap_mb
+                    return train.torch.prepare_model(model, ddp_kwargs=kwargs)
+
+                def _save(self, *args, **kwargs):
+                    # Workaround for RayTrainingArguments not being
+                    # pickleable due to it being defined in a local
+                    # scope
+                    self.args.__class__ = base_training_arguments_class
+                    ret = super()._save(*args, **kwargs)
+                    self.args.__class__ = RayTrainingArguments
+                    return ret
+
+            trainer.__class__ = RayTrainer
+            trainer.args.__class__ = RayTrainingArguments
+            trainer.args.no_cuda = not torch.cuda.is_available()
+            trainer.args.save_on_each_node = True
+
+            # ensure no HF logging callbacks are added
+            # aside from doubling functionality with our callbacks,
+            # the Wandb callbacks causes training to freeze
+            integration_callbacks = (
+                transformers.trainer.get_reporting_integration_callbacks(
+                    trainer.args.report_to
+                )
+            )
+            for callback in integration_callbacks:
+                trainer.pop_callback(callback)
```

Comment:
This is defined by the user in the factory function. I don't know how we can ensure that it is none without either doing this or monkey patching.

Reply:
Could we do something like this?

```
if trainer.args.report_to != "none":
   raise ValueError(...)
```

Reply:
`report_to` gets replaced with a list of various callbacks during `TrainingArguments` init:
```
        if self.report_to is None:
            logger.info(
                "The default value for the training argument `--report_to` will change in v5 (from all installed "
                "integrations to none). In v5, you will need to use `--report_to all` to get the same behavior as "
                "now. You should start updating your code and make this info disappear :-)."
            )
            self.report_to = "all"
        if self.report_to == "all" or self.report_to == ["all"]:
            # Import at runtime to avoid a circular import.
            from .integrations import get_available_reporting_integrations

            self.report_to = get_available_reporting_integrations()
        elif self.report_to == "none" or self.report_to == ["none"]:
            self.report_to = []
        elif not isinstance(self.report_to, list):
            self.report_to = [self.report_to]
            ```

But I suppose we can just check for an empty list then. It will be one more thing we will require the user to set themselves, though. I think we would want to avoid that as much as possible, for better UX.